File: android/core/protocol/src/main/kotlin/app/hapi/protocol/wire/SyncEvent.kt
package app.hapi.protocol.wire

import kotlinx.serialization.DeserializationStrategy
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

/**
 * The SSE event union (`SyncEventSchema`, `shared/src/schemas.ts`;
 * `docs/api/client-contract/sse.md` — 13 types, discriminated on `type`).
 * Decode with [SyncEvents.parse]; there is no polymorphic serializer because
 * frames arrive one JSON document per SSE `data:` line and unknown /
 * malformed frames must degrade to [Unknown] instead of throwing.
 *
 * `session-updated` / `session-added` / `machine-updated` keep their `data`
 * raw: resolving full-object vs patch is the caller's job (full [Session] /
 * [Machine] via [HapiJson], patch via [SessionPatches.parse] /
 * [MachinePatches.parse], anything else → REST refetch fallback).
 */
sealed interface SyncEvent {
    val namespace: String?

    @Serializable
    data class SessionAdded(
        override val namespace: String? = null,
        val sessionId: String,
        val data: JsonElement? = null,
    ) : SyncEvent

    @Serializable
    data class SessionUpdated(
        override val namespace: String? = null,
        val sessionId: String,
        /** Full `Session` or `SessionPatch`; absent/unparseable ⇒ refetch. */
        val data: JsonElement? = null,
    ) : SyncEvent

    @Serializable
    data class SessionRemoved(
        override val namespace: String? = null,
        val sessionId: String,
    ) : SyncEvent

    @Serializable
    data class MessageReceived(
        override val namespace: String? = null,
        val sessionId: String,
        val message: DecryptedMessage,
    ) : SyncEvent

    /** History changed structurally (rewind/fork/import/clear): drop the window, tail-sync. */
    @Serializable
    data class MessagesInvalidated(
        override val namespace: String? = null,
        val sessionId: String,
    ) : SyncEvent

    @Serializable
    data class ScheduledMatured(
        override val namespace: String? = null,
        val sessionId: String,
    ) : SyncEvent

    @Serializable
    data class SessionEnded(
        override val namespace: String? = null,
        val sessionId: String,
        /** `'completed' | 'terminated' | 'error' | 'handoff' | 'cleared'`. */
        val reason: String? = null,
    ) : SyncEvent

    /**
     * Not `@Serializable`: built by hand in [SyncEvents.parse] because `data`
     * is tri-state — `Present(JsonNull)` = machine **removed**,
     * `Present(object)` = full `Machine` or `MachinePatch`, [OptionalField.Absent]
     * = refetch machines (`docs/api/client-contract/sse.md`). A nullable
     * `JsonElement?` property would collapse the first and last cases.
     */
    data class MachineUpdated(
        override val namespace: String? = null,
        val machineId: String,
        val data: OptionalField<JsonElement> = OptionalField.Absent,
    ) : SyncEvent

    @Serializable
    data class Toast(
        override val namespace: String? = null,
        val data: Data,
    ) : SyncEvent {
        @Serializable
        data class Data(
            val title: String,
            val body: String,
            val sessionId: String,
            val url: String,
        )
    }

    /** The agent consumed queued user messages: stamp `invokedAt`, leave the queued bar. */
    @Serializable
    data class MessagesConsumed(
        override val namespace: String? = null,
        val sessionId: String,
        val localIds: List<String>,
        val invokedAt: Long,
    ) : SyncEvent

    @Serializable
    data class MessagesIndeterminate(
        override val namespace: String? = null,
        val sessionId: String,
        val localIds: List<String>,
    ) : SyncEvent

    @Serializable
    data class MessagesRequeued(
        override val namespace: String? = null,
        val sessionId: String,
        val localIds: List<String>,
    ) : SyncEvent

    @Serializable
    data class MessageCancelled(
        override val namespace: String? = null,
        val sessionId: String,
        val messageId: String,
        val localId: String? = null,
    ) : SyncEvent

    /** Staleness-watchdog food; carries no SSE `id`. */
    @Serializable
    data class Heartbeat(
        override val namespace: String? = null,
        val data: Data? = null,
    ) : SyncEvent {
        @Serializable
        data class Data(
            val timestamp: Long,
        )
    }

    /** Subscribe handshake; carries no SSE `id`. */
    @Serializable
    data class ConnectionChanged(
        override val namespace: String? = null,
        val data: Data? = null,
    ) : SyncEvent {
        @Serializable
        data class Data(
            val status: String,
            /** Needed for `POST /api/visibility` reporting. */
            val subscriptionId: String? = null,
            /** `'ok'` = replay is complete, skip resync; `'gap'`/absent = full REST resync. */
            val resume: String? = null,
        )
    }

    /**
     * Unrecognized `type`, or a known `type` whose payload failed to decode
     * (the web reference likewise drops frames failing `safeParse`). Callers
     * treat it as a no-op; [raw] is retained for diagnostics.
     */
    data class Unknown(
        val type: String?,
        val raw: JsonObject? = null,
    ) : SyncEvent {
        override val namespace: String? get() = raw?.get("namespace").stringOrNull
    }
}

object SyncEvents {
    /**
     * Total decode of one SSE `data:` payload. Never throws: invalid JSON,
     * non-object frames, unknown `type`s, and malformed known-type payloads
     * all become [SyncEvent.Unknown].
     */
    fun parse(json: String): SyncEvent {
        val element = try {
            HapiJson.parseToJsonElement(json)
        } catch (_: Exception) {
            return SyncEvent.Unknown(type = null, raw = null)
        }
        return parse(element)
    }

    fun parse(element: JsonElement): SyncEvent {
        val obj = element as? JsonObject ?: return SyncEvent.Unknown(type = null, raw = null)
        val type = obj["type"].stringOrNull ?: return SyncEvent.Unknown(type = null, raw = obj)
        val strategy: DeserializationStrategy<SyncEvent> = when (type) {
            "session-added" -> SyncEvent.SessionAdded.serializer()
            "session-updated" -> SyncEvent.SessionUpdated.serializer()
            "session-removed" -> SyncEvent.SessionRemoved.serializer()
            "message-received" -> SyncEvent.MessageReceived.serializer()
            "messages-invalidated" -> SyncEvent.MessagesInvalidated.serializer()
            "scheduled-matured" -> SyncEvent.ScheduledMatured.serializer()
            "session-ended" -> SyncEvent.SessionEnded.serializer()
            "machine-updated" -> return parseMachineUpdated(obj)
            "toast" -> SyncEvent.Toast.serializer()
            "messages-consumed" -> SyncEvent.MessagesConsumed.serializer()
            "messages-indeterminate" -> SyncEvent.MessagesIndeterminate.serializer()
            "messages-requeued" -> SyncEvent.MessagesRequeued.serializer()
            "message-cancelled" -> SyncEvent.MessageCancelled.serializer()
            "heartbeat" -> SyncEvent.Heartbeat.serializer()
            "connection-changed" -> SyncEvent.ConnectionChanged.serializer()
            else -> return SyncEvent.Unknown(type = type, raw = obj)
        }
        return try {
            HapiJson.decodeFromJsonElement(strategy, obj)
        } catch (_: Exception) {
            SyncEvent.Unknown(type = type, raw = obj)
        }
    }

    private fun parseMachineUpdated(obj: JsonObject): SyncEvent {
        val machineId = obj["machineId"].stringOrNull
            ?: return SyncEvent.Unknown(type = "machine-updated", raw = obj)
        return SyncEvent.MachineUpdated(
            namespace = obj["namespace"].stringOrNull,
            machineId = machineId,
            data = if (obj.containsKey("data")) {
                OptionalField.Present(obj.getValue("data"))
            } else {
                OptionalField.Absent
            },
        )
    }
}

File: ios/Packages/HapiKit/Sources/HapiProtocol/Models/SyncEvent.swift
import Foundation

/// Why a session ended (`session-ended` event).
public enum SessionEndReason: String, Codable, Sendable {
    case completed
    case terminated
    case error
    case handoff
    case cleared
}

/// Payload of a `toast` event (visibility-targeted in-app notification).
public struct ToastPayload: Codable, Equatable, Sendable {
    public var title: String
    public var body: String
    public var sessionId: String
    public var url: String

    public init(title: String, body: String, sessionId: String, url: String) {
        self.title = title
        self.body = body
        self.sessionId = sessionId
        self.url = url
    }
}

/// Reconnect verdict inside `connection-changed`.
///
/// `ok` = the hub replays every missed event right after the handshake, so
/// the client may skip its REST resync. `gap` — or an absent/unknown verdict
/// — means a full refetch is required.
public enum ResumeVerdict: String, Codable, Sendable {
    case ok
    case gap
}

/// Payload of the `connection-changed` handshake frame.
public struct ConnectionChangedPayload: Codable, Equatable, Sendable {
    public var status: String
    /// Needed for `POST /api/visibility`; new on every reconnect.
    public var subscriptionId: String?
    /// Decoded leniently: an unknown wire verdict becomes `nil`, which
    /// clients must treat as `gap` (the contract's "absence" rule).
    public var resume: ResumeVerdict?

    public init(status: String, subscriptionId: String? = nil, resume: ResumeVerdict? = nil) {
        self.status = status
        self.subscriptionId = subscriptionId
        self.resume = resume
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        status = try container.decode(String.self, forKey: .status)
        subscriptionId = try container.decodeIfPresent(String.self, forKey: .subscriptionId)
        if let raw = try container.decodeIfPresent(String.self, forKey: .resume) {
            resume = ResumeVerdict(rawValue: raw)
        } else {
            resume = nil
        }
    }
}

/// The `session-updated` (and `session-added`) data union:
/// a full `Session`, a strict `SessionPatch`, or something this client cannot
/// interpret (the caller must fall back to a REST refetch, mirroring the web
/// client's behavior when `SessionUpdatedDataSchema` fails to parse).
public enum SessionUpdatedData: Equatable, Sendable {
    case session(Session)
    case patch(SessionPatch)
    case unrecognized(JSONValue)
}

extension SessionUpdatedData: Decodable {
    public init(from decoder: Decoder) throws {
        // Same order as `z.union([SessionSchema, SessionPatchSchema])`.
        if let session = try? Session(from: decoder) {
            self = .session(session)
            return
        }
        if let patch = try? SessionPatch(from: decoder) {
            self = .patch(patch)
            return
        }
        self = .unrecognized(try JSONValue(from: decoder))
    }
}

/// The `machine-updated` data union: full `Machine`, strict `MachinePatch`,
/// explicit `null` (machine removed), or unrecognized (refetch machines).
public enum MachineUpdatedData: Equatable, Sendable {
    case machine(Machine)
    case patch(MachinePatch)
    case removed
    case unrecognized(JSONValue)
}

extension MachineUpdatedData: Decodable {
    public init(from decoder: Decoder) throws {
        if let single = try? decoder.singleValueContainer(), single.decodeNil() {
            self = .removed
            return
        }
        if let machine = try? Machine(from: decoder) {
            self = .machine(machine)
            return
        }
        if let patch = try? MachinePatch(from: decoder) {
            self = .patch(patch)
            return
        }
        self = .unrecognized(try JSONValue(from: decoder))
    }
}

/// One frame of the `GET /api/events` SSE stream.
///
/// Mirrors `SyncEventSchema` (`shared/src/schemas.ts:535-605`) and
/// `docs/api/client-contract/sse.md`. Every case carries the optional
/// `namespace`. Unknown event types decode to `.unknown` — never throw —
/// so a hub newer than this client degrades gracefully; known types with
/// malformed payloads throw, mirroring a zod parse failure.
public enum SyncEvent: Equatable, Sendable {
    /// Handle exactly like `session-updated` (the web client shares the
    /// branch): `.session` upserts, anything else falls back to a refetch.
    case sessionAdded(namespace: String?, sessionId: String, data: SessionUpdatedData?)
    case sessionUpdated(namespace: String?, sessionId: String, data: SessionUpdatedData?)
    case sessionRemoved(namespace: String?, sessionId: String)
    case messageReceived(namespace: String?, sessionId: String, message: DecryptedMessage)
    /// History changed structurally (rewind/fork/import/clear): discard the
    /// window and tail-sync.
    case messagesInvalidated(namespace: String?, sessionId: String)
    /// A scheduled message became due and was handed to the agent.
    case scheduledMatured(namespace: String?, sessionId: String)
    /// `reason` is `nil` when absent or unknown on the wire.
    case sessionEnded(namespace: String?, sessionId: String, reason: SessionEndReason?)
    /// `data` is `nil` when the wire field is absent (refetch machines).
    case machineUpdated(namespace: String?, machineId: String, data: MachineUpdatedData?)
    case toast(namespace: String?, data: ToastPayload)
    /// Queued user messages were consumed: stamp `invokedAt` on the rows.
    case messagesConsumed(namespace: String?, sessionId: String, localIds: [String], invokedAt: Int)
    case messagesIndeterminate(namespace: String?, sessionId: String, localIds: [String])
    case messagesRequeued(namespace: String?, sessionId: String, localIds: [String])
    case messageCancelled(namespace: String?, sessionId: String, messageId: String, localId: String?)
    /// Feeds the staleness watchdog only. Carries no SSE `id`.
    case heartbeat(namespace: String?, timestamp: Int?)
    /// Subscription handshake. Carries no SSE `id`.
    case connectionChanged(namespace: String?, data: ConnectionChangedPayload?)
    /// Any `type` this client does not know. Ignore.
    case unknown(type: String, namespace: String?)
}

extension SyncEvent: Decodable {
    private enum CodingKeys: String, CodingKey {
        case type
        case namespace
        case sessionId
        case machineId
        case data
        case message
        case reason
        case localIds
        case invokedAt
        case messageId
        case localId
    }

    private struct HeartbeatData: Decodable {
        let timestamp: Int
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(String.self, forKey: .type)
        let namespace = try container.decodeIfPresent(String.self, forKey: .namespace)
        switch type {
        case "session-added":
            self = .sessionAdded(
                namespace: namespace,
                sessionId: try container.decode(String.self, forKey: .sessionId),
                data: try container.decodeIfPresent(SessionUpdatedData.self, forKey: .data)
            )
        case "session-updated":
            self = .sessionUpdated(
                namespace: namespace,
                sessionId: try container.decode(String.self, forKey: .sessionId),
                data: try container.decodeIfPresent(SessionUpdatedData.self, forKey: .data)
            )
        case "session-removed":
            self = .sessionRemoved(
                namespace: namespace,
                sessionId: try container.decode(String.self, forKey: .sessionId)
            )
        case "message-received":
            self = .messageReceived(
                namespace: namespace,
                sessionId: try container.decode(String.self, forKey: .sessionId),
                message: try container.decode(DecryptedMessage.self, forKey: .message)
            )
        case "messages-invalidated":
            self = .messagesInvalidated(
                namespace: namespace,
                sessionId: try container.decode(String.self, forKey: .sessionId)
            )
        case "scheduled-matured":
            self = .scheduledMatured(
                namespace: namespace,
                sessionId: try container.decode(String.self, forKey: .sessionId)
            )
        case "session-ended":
            let reason = try container.decodeIfPresent(String.self, forKey: .reason)
                .flatMap(SessionEndReason.init(rawValue:))
            self = .sessionEnded(
                namespace: namespace,
                sessionId: try container.decode(String.self, forKey: .sessionId),
                reason: reason
            )
        case "machine-updated":
            let machineId = try container.decode(String.self, forKey: .machineId)
            // `data: null` means "machine removed" and must stay distinct
            // from an absent `data` (refetch); `decodeIfPresent` would
            // conflate the two, so probe presence and null by hand.
            let data: MachineUpdatedData?
            if container.contains(.data) {
                if try container.decodeNil(forKey: .data) {
                    data = .removed
                } else {
                    data = try container.decode(MachineUpdatedData.self, forKey: .data)
                }
            } else {
                data = nil
            }
            self = .machineUpdated(namespace: namespace, machineId: machineId, data: data)
        case "toast":
            self = .toast(
                namespace: namespace,
                data: try container.decode(ToastPayload.self, forKey: .data)
            )
        case "messages-consumed":
            self = .messagesConsumed(
                namespace: namespace,
                sessionId: try container.decode(String.self, forKey: .sessionId),
                localIds: try container.decode([String].self, forKey: .localIds),
                invokedAt: try container.decode(Int.self, forKey: .invokedAt)
            )
        case "messages-indeterminate":
            self = .messagesIndeterminate(
                namespace: namespace,
                sessionId: try container.decode(String.self, forKey: .sessionId),
                localIds: try container.decode([String].self, forKey: .localIds)
            )
        case "messages-requeued":
            self = .messagesRequeued(
                namespace: namespace,
                sessionId: try container.decode(String.self, forKey: .sessionId),
                localIds: try container.decode([String].self, forKey: .localIds)
            )
        case "message-cancelled":
            self = .messageCancelled(
                namespace: namespace,
                sessionId: try container.decode(String.self, forKey: .sessionId),
                messageId: try container.decode(String.self, forKey: .messageId),
                localId: try container.decodeIfPresent(String.self, forKey: .localId)
            )
        case "heartbeat":
            self = .heartbeat(
                namespace: namespace,
                timestamp: try container.decodeIfPresent(HeartbeatData.self, forKey: .data)?.timestamp
            )
        case "connection-changed":
            self = .connectionChanged(
                namespace: namespace,
                data: try container.decodeIfPresent(ConnectionChangedPayload.self, forKey: .data)
            )
        default:
            self = .unknown(type: type, namespace: namespace)
        }
    }
}
